### PR TITLE
Q2 2022

### PR DESCRIPTION
**Pull request by bethanne-card**
Added in patent_crosswalk query into the pipeline. I also added in some QA work to move organization_name from first name or last name and a QA check.

### CODE REVIEW

**Comment by jbutler-air** [comment]
A couple notes:
on accuracy/robustness:
* Fix_rawassignee_wrong_org corrects for both cases where an assignee has first name but not last and last name but not first, but test_rawassignee_org only checks for first name with no last name. We can add an ‘OR’ condition to the test_rawassignee_org query (example 1 below)
* fix_rawassignee_wrong_org checks for a string match of keywords in the last name, but not in the first name; is this an intentional difference, or should the function the same?
* fix_rawassignee_wrong_org changes all ‘type’ values to 2 when it moves a name from first or last name to org name, but doesn’t check that the organizations are US-based (international orgs get type=3). We can either incorporate a ‘CASE WHEN’ phrase into the ‘SET’ clause, or we can run two otherwise identical queries where one makes the name change and sets type=2 where country=US and the other makes the name change and sets type=3 where country<>US. (Example 2 below)

on efficiency:
* We should be able to get around the temporary tables in fix_rawassignee_wrong_org - I think we can just do “SET organization = name_{first/last}, type = {2/3}, name_{first/last} = NULL” - my understanding is these will resolve in order and set org to name before changing name to null. (Example 2 below) 

SELECT COUNT(*)
FROM rawassignee
WHERE (name_first IS NOT NULL AND name_last IS NULL)
OR (name_first IS NULL AND name_last IS NOT NULL)


UPDATE rawassignee
SET organization = name_{first/last}, name_{first/last} = NULL, type = (CASE WHEN country = ‘US’ THEN ‘2’ ELSE ‘3’)
WHERE where name_{first/last} IS NOT NULL 
AND name_{last/first} IS NULL
AND name_{first/last} REGEXP ‘{keywords}’

**Comment by bethanne-card** [comment]
Thanks for looking into this Jack!

So there are logic differences between the two cases for a reason. When there is first_name and no last name, we assume that all those entries are an organization and can move first_name entries to the organization column and then easily check for it. In the case of last name with no first name, if you look at the data, some of the names look like an actual last name while some of the rows are organizations. I did my best to subset the organizations and move those rows from last_name to organization. We can not write a simple test since not all of the rows are able to be moved with certainty. Come to think of it, we can at least make sure that all the rows with organization-related substrings were moved over.

Do you want to add that test or I can?

Thanks again! Happy Friday 😄
BA
________________________________

File: lib/utilities.py [hunk 1]
```diff
@@ -294,12 +294,17 @@ def download(url, filepath):
...
-        content_length = int(r.headers.get('content-length'))
-        for chunk in progress.bar(r.iter_content(chunk_size=1024),
-                                  expected_size=(content_length / 1024) + 1):
-            if chunk:
-                f.write(chunk)
-                f.flush()
+        # print(r.headers.get('content-length'))
+        content_length = r.headers.get('content-length')
+        if not content_length:
+            print("/tNo Content Length Attached")
```
Comment (sarvothaman) [inline lib/utilities.py:300]: What's with the `/t`? do you mean `\t`?

File: airflow/dags/reporting_database_generation/generate-reporting-db.py [hunk 2]
```diff
@@ -577,11 +582,23 @@ class SQLTemplatedPythonOperator(PythonOperator):
 #     params=database_name_config
 # )
 
+# OTHER MISC TASKS TO BE RUN
+create_granted_patent_crosswalk = PythonOperator(task_id='create_granted_patent_crosswalk',
```
Comment (sarvothaman) [inline airflow/dags/reporting_database_generation/generate-reporting-db.py:586]: Are there no other DAGs that's is better suited for this task? Technically, this belong to "raw database" creation. Further the bulk download process is understood as dependent on raw database creation and considered ready to start as soon raw database is done. Having this task here may introduce inconsistencies

File: QA/generic_tests.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    rows = engine.execute(patent_cq_query).fetchone()[0]
+    distinct_vi = engine.execute(patent_cq_query).fetchone()[1]
+    query_end_time = time()
+    print("This query took:", query_end_time - query_start_time, "seconds")
+    print(f"{table} HAS {distinct_vi} UNIQUE WEEKS OF DATA")
+    if rows == 0:
+        raise Exception(f"{db}.{table} WAS NOT UPDATED FOR THE CURRENT TIMEFRAME")
+
+
+def update_to_granular_version_indicator(table, db):
```
Comment (sarvothaman) [inline QA/generic_tests.py:33]: I thought i saw a version of this already defined elsewhere? In any case, this function and the `update_version_indicator` should move out of QA folder. (Perhaps into lib/utilities.py?)

**Review by sarvothaman — CHANGES_REQUESTED** [review]

File: updater/xml_to_sql/post_processing.py [hunk 4]
```diff
@@ -246,11 +246,76 @@ def trim_rawassignee(config):
         (name_last IS NULL) AND
         (organization IS NULL);""")
 
+def fix_rawassignee_wrong_org(config):
```
Comment (sarvothaman) [inline updater/xml_to_sql/post_processing.py:249]: Is there a risk of false positives here? i.e. records we thought were organization was in reality just someone with just first name. ( https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/ ) . If the existing data supports the theory that in _patent_ data, this seems to be the case most/all of the time, then I would suggest creating "audit" records to keep track of which records were changed in post processing.

**Comment by sarvothaman** [comment]
@bethanne-card  I don't have anything else. Feel free to merge when you are ready

**Comment by bethanne-card** [comment]
We do keep a record in a table called temp_rawassignee_org_fixes_nf

or do you mean we should have more of a generalized table that keeps track of records changed from any table?

________________________________

**Comment by bethanne-card** [comment]
I meant we’ll need more info (what- rawassignee, where – pgpubs,  when – date of change , who – update pipeline) to be stored in that table if it’s going to be automated. The table name itself can answer some of that info (for eg. rawasisgnee_org tells us its rawassignee).

Also don’t temp_ tables trigger a QA failure? Perhaps a different prefix make sense?

Thanks,
Sarvo
From: Card, Beth Anne ***@***.***>
Date: Friday, May 6, 2022 at 1:04 PM
To: PatentsView/PatentsView-DB ***@***.***>, Madhavan, Sarvothaman ***@***.***>
Subject: Re: [PatentsView/PatentsView-DB] Q2 2022 (PR #110)
We do keep a record in a table called temp_rawassignee_org_fixes_nf

or do you mean we should have more of a generalized table that keeps track of records changed from any table?

________________________________

**Pull request merged**